### PR TITLE
Add owner field to assets

### PR DESCRIPTION
**Pull request by jgOhYeah**
Hello,
This is a cleaned up branch. The previous pull request with old branch is https://github.com/farmOS/farmOS/pull/532.

This pull request adds an owner field to assets, based off the assign to functionality of logs.
For discussion and further details, please see here: https://farmos.discourse.group/t/is-it-possible-to-use-custom-fields-and-filtering-in-asset-views-and-filter-by-a-range-of-birth-dates/1139

### CODE REVIEW

File: modules/core/owner/farm_owner.module
Comment (paul121) [inline modules/core/owner/farm_owner.module]: Same, maybe update this field description as well.

File: modules/core/owner/farm_owner.post_update.php
Comment (paul121) [inline modules/core/owner/farm_owner.post_update.php]: Should we change this field description since assets are not "tasks" ? Is there more general language we could use here?

File: modules/core/owner/config/schema/farm_owner.schema.yml
Comment (paul121) [inline modules/core/owner/config/schema/farm_owner.schema.yml]: I think we need to implement a hook to update existing schema on sites if we change this here. I haven't done this before but think I have seen it in other projects, maybe the `log` module most recently?

File: modules/core/owner/config/optional/system.action.asset_assign_action.yml
Comment (paul121) [inline modules/core/owner/config/optional/system.action.asset_assign_action.yml]: While "Assign log" makes sense, I'm not sure that "Assign asset" makes as much sense. Maybe "Assign owner"?

File: modules/core/owner/src/Form/AssignActionForm.php
Comment (paul121) [inline modules/core/owner/src/Form/AssignActionForm.php]: Instead of the switch statement I think we could use single `t()` function but insert `asset` or `log` in as a parameter. This should work for both the title, users description, and operation description because all of the strings here are nearly identical! The benefit is that it is less strings to translate.

Although if we did want to make major changes to the strings, this switch statement looks great.

File: modules/core/ui/views/config/install/views.view.farm_asset.yml
Comment (paul121) [inline modules/core/ui/views/config/install/views.view.farm_asset.yml]: "Assigned to" or "owner" ?

File: modules/core/owner/src/Plugin/Action/AssetAssign.php
Comment (paul121) [inline modules/core/owner/src/Plugin/Action/AssetAssign.php:15]: This is great, love the idea to use the `AssignBase` class.

**Review by paul121 — DISMISSED** [review]
This looks pretty good @jgOhYeah !

Generally I'm curious about the language of this field, "Assigned assets" or "Owned assets", etc. It comes up a few places here. Apologies if that was already "decided" elsewhere, just something that came to mind.

**Comment by jgOhYeah** [comment]
Thanks for looking at it. The 'assign asset' language was mainly because I was trying to keep the wording fairly similar between logs and assets - it can be changed. Would 'Owner' be the more correct label in this case? As a side note, [this Drupal issue](https://www.drupal.org/project/farm/issues/2400065) discusses some of the various interpretations of 'ownership' of assets, which may affect what the field gets called.

File: modules/core/owner/src/Plugin/Action/AssetAssign.php
Comment (jgOhYeah) [reply modules/core/owner/src/Plugin/Action/AssetAssign.php:15]: Thanks to @mstenta for the suggestion.

File: modules/core/owner/src/Form/AssignActionForm.php
Comment (jgOhYeah) [reply modules/core/owner/src/Form/AssignActionForm.php]: I did have the parameter substitution set up like you suggest at one point, however the php codesniffer tests didn't really like the idea of putting variable substitutions in `t()` functions. I'm not a language expert, but I recall that some languages have words in different orders to others, which I'm not sure how well the variable substitution would handle. I can change it back if this isn't an issue.

File: modules/core/owner/config/schema/farm_owner.schema.yml
Comment (jgOhYeah) [reply modules/core/owner/config/schema/farm_owner.schema.yml]: Would it be possible to link to a previous example if you know roughly where one is please? I've had a look for update hooks and post update hooks in the `/modules/log/` and `/modules/core/log` folders to base this off, but haven't come across any. I could be missing something subtle (or pretty darn obvious :)), but the assign log action still seems to behave alright on an existing site, even without this update hook.

**Comment by mstenta** [comment]
Just thought about this again while adding some records to my farmOS recently... sorry for the long delay - been very busy with things.

I wonder: should we change the label of the field on Logs to "Owner" as well? Perhaps that's what you were implying too @jgOhYeah:

> Would 'Owner' be the more correct label in this case?

It is more general than "Assigned to", which means it works in more scenarios. "Assigned to" is nice because it's easy to understand *in the context of delegated work assignments*, so "Owner" is a bit of a step down in that context, but maybe the tradeoff is acceptable? It generalizes the meaning a bit, which allows it to be more flexible depending on context and convention.

There are contexts where "Assigned to" feels weird - for example if you are a one-person operation. I suppose "Owner" is odd in that context as well, though. The idea is "who performed this task?" What is the best label for that? "Owner" is the right *idea* I still believe (and therefore machine name). So perhaps it's the best *label* as well.

**Comment by jgOhYeah** [comment]
> Perhaps that's what you were implying too @jgOhYeah:
I was originally confirming that 'Owner' was the correct replacement for 'Assigned to' for assets, but it wasn't particularly clearly worded and looking back again see that or something similar is what @paul121 was suggesting in the first place.

I don't really have a preference as to the naming. The only risk with changing the log field label is consistency for those who are used to the old label, but this isn't a major change.

I agree that depending on the situation, 'Assigned to' vs 'Owner' have their own advantages and disadvantages. For our installation it can sometimes be a presumed who did it as well :).

**Comment by mstenta** [comment]
Finally got a chance to dig in and review/test this! Thanks again @jgOhYeah and I appreciate your patience. :-)

> I wonder: should we change the label of the field on Logs to "Owner" as well?

I think I am leaning more towards this. It would allow us to keep a lot more of these labels and strings consistent, which has some benefits (eg: less code, easier to maintain translations).

I rebased the working branch, squashed the codesniffer commits into the first commit, and then added a new commit that updates the language to this effect.

I used "owner" in noun contexts, and "assign" in verb contexts, roughly speaking. It still uses "Assigned assets" and "Assigned logs" in the Views of records owned by users, but I think that's OK.

In this sense, we can think of it as "_record_ ownership" - as in "who is in charge of this record" (whatever that is taken to mean). Some may use this to mean "task assignment" (in the case of logs), or "who is in charge of this" (in the case of logs/assets), others may use it to mean "the legal owner" (of assets).

Lastly, I want to raise one last idea/question: should we auto-assign asset owners in the same way we auto-assign log owners (when a log is saved, if no owner is assigned, it assigns it to the current user)? I actually went ahead and implemented this (which required also opening this MR: https://www.drupal.org/project/farm/issues/3306227) - but then in testing it I realized this means it's going to add "Owner" to all existing assets - and maybe we don't really want that. Maybe we should leave it up to the user to decide how ownership is assigned for assets. Although it feels weird for logs and assets to behave differently in this aspect, doesn't it? I kept this change out of the branch for now, but I'd like to hear what others think. We could also punt on this decision and open a followup issue for consideration (I can add my commit there).

I feel like this is ready to merge, pending thoughts/review of all of the above.

**Comment by mstenta** [comment]
Meant to respond to this @paul121:

> I think we need to implement a hook to update existing schema on sites if we change this here.

I'm not sure anything is needed for schema changes, is it? Maybe I'm not understanding.

**Comment by symbioquine** [comment]
> it's going to add "Owner" to all existing assets - and maybe we don't really want that. Maybe we should leave it up to the user to decide how ownership is assigned for assets. Although it feels weird for logs and assets to behave differently in this aspect, doesn't it? I kept this change out of the branch for now, but I'd like to hear what others think.

My gut feeling is that it's a little odd to make the logic assume the asset creator is the asset owner.

The best rational for that stance I can come up with is that the lifecycle of assets and logs is often quite different - many assets are perennial or are long-lived equipment/land-features/etc. Contrasted with logs where I believe the default of self-assigning logs is pretty sane since the person who created the log presumably has logical ownership of the task - at least insofar as capturing it as a log goes - and may already have completed it themselves.

Maybe it would be better to provide a "default interpretation" of the no-owner state for assets (and make that the default) which somehow describes them as "belonging to the farm as a whole"?

**Comment by mstenta** [comment]
I like that reasoning @symbioquine - and leaving this as-is means that "Owner" will not appear on assets unless it is filled in. (There is still an "Owner" column" in the Views, which will be empty, but I think that's OK for now.)

**Comment by jgOhYeah** [comment]
Thanks for fixing this up. I would be inclined to agree with @symbioquine 's reasoning at this stage. In our case we have certain assets like implements that require more research on who exactly owns them or they don't really have an owner, so being able to leave than field blank by default would be useful.

**Comment by mstenta** [comment]
Ah hmm tests are failing:

```
Drupal\Core\Config\Schema\SchemaIncompleteException: Schema errors for system.action.asset_assign_action with the following errors: system.action.asset_assign_action:configuration missing schema
```

**Comment by mstenta** [comment]
> Ah hmm tests are failing:

I pushed a commit that attempts to fix this... waiting for results from GitHub Actions...

https://github.com/mstenta/farmOS/actions/runs/2934888499

**Comment by mstenta** [comment]
Getting a different schema error now, related to the changes to Views config:

`Drupal\Core\Config\Schema\SchemaIncompleteException: Schema errors for views.view.farm_people with the following errors: views.view.farm_people:display.default.display_options.nothing missing schema`

Will need to debug that another time...

**Comment by mstenta** [comment]
I figured out the other schema error. @jgOhYeah it appears you may have copied and pasted a chunk of code into the `views.view.farm_people.yml` file, does that sound correct? There were two chunks added to that file, but the second one was invalid. Deleting it allowed the tests to pass. It does not affect the intention of the change, so the View still has the new "Assigned assets" column. Let me know if you have any concerns, but I think that explains what happened.

I pushed a fixup commit for this, but I will squash it into the first commit of the branch before merging. I just kept it separate for review.

**Comment by jgOhYeah** [comment]
> I figured out the other schema error. @jgOhYeah it appears you may have copied and pasted a chunk of code into the `views.view.farm_people.yml` file, does that sound correct? There were two chunks added to that file, but the second one was invalid. Deleting it allowed the tests to pass. It does not affect the intention of the change, so the View still has the new "Assigned assets" column. Let me know if you have any concerns, but I think that explains what happened.
> 
> I pushed a fixup commit for this, but I will squash it into the first commit of the branch before merging. I just kept it separate for review.

I think that must have happened during my attempt at unscrambling some of the merges I'd done and making a fresh branch. I recall that involved a couple of manual merges and something obviously went wrong. Thanks for fixing it, sorry it caused so much bother.

**Comment by mstenta** [comment]
No worries! Easy fix. :-)

I think this is ready for final review and merge now. All tests are passing.

**Comment by mstenta** [comment]
(Just merged my fixup into the initial commit and rebased onto latest `2.x` to resolve `CHANGELOG.md` conflict.)

**Comment by mstenta** [comment]
Rebased and added one more commit to document the new owner field in the user guide and data model docs.

If tests pass I'll merge this.

**Comment by mstenta** [comment]
Merged! At long last!!!

Thanks again @jgOhYeah! :-D :tada:

**Comment by jgOhYeah** [comment]
Thanks for all the effort you have put into this @mstenta and others. This has / will make farmOS much more useful for us and hopefully will for others.

**Pull request merged**
